### ISSUE TITLE
Http/2 Support ?

### ISSUE THREAD

**Issue by vefaliahmet** [EasyEngine/easyengine#636]
Nginx 1.9.5 is supporting http/2 protocol. But EasyEngine's nginx version is 1.8. Chrome will remove SPDY support in early 2016. Is there any plan about HTTP/2 support?

**Comment by ddelaey** [comment]
+1 for HTTP/2 support

**Comment by ldashdixon** [comment]
+1 if a petition is needed. :)

**Comment by MattyLB** [comment]
+5

**Comment by cm896** [comment]
+1

**Comment by puikinsh** [comment]
+1

**Comment by MiteshShah** [comment]
http://nginx.org/en/download.html

As of now nginx stable version is 1.8.0
When NGINX merge mainline version in to stable version I guess @gau1991 update the nginx package so we have http/2 support.

**Comment by codxse** [comment]
yup, agreed. Keep the stable version or make it on other branch. Im using easyengine on production site. It's too risk if a failure happen, sometime unpredictable error happen on unstable version.

**Comment by gau1991** [comment]
Hello all,
Nginx 1.9.5 is not stable yet. We will update Nginx package in our repo once mainline version is merged into stable.

**Comment by ichpen** [comment]
I don't understand the reasoning behind this as the NGINX team themselves recommend using mainline versions where possible as it has the latest features and most up-to-date bug fixes vs. older stable. The only possible issue is third party modules which in most cases can easily be traced. @gau1991

**Comment by ichpen** [comment]
@gau1991 Please refer to this link for more info on how nginx versioning works:
https://www.nginx.com/blog/nginx-1-6-1-7-released/

In particular this snippet may be of interest:
"Which Version Should I Use?

We recommend that in general you deploy the NGINX mainline branch at all times. The main reason to use the stable branch is that you are concerned about possible impacts of new features, such as incompatibility with third-party modules or the inadvertent introduction of bugs in new features."

**Comment by w33zy** [comment]
I agree with @ichpen considering that ee was released with support for HHVM and Redis when they were well into their "experimental" stages. Even now HHVM isn't perfect for production use with WP, as it still eats memory, albeit a bit more slowly.

Adding support for Nginx mainline should be a long stretch for the team.

Plus looking the rtCamp business, I know you guys are probably using v1.9.5 on your production servers. I know it would be a plus, telling potential clients about your HTTP/2 server support.

**Comment by MattyLB** [comment]
@w33zy - I agree with every point of that. I've gone as far as to remove
the hhvm and redis stacks because they're problematic for my WP sites and
I'd rather not run them on my server. With HTTP/2 - its benefits for page
loading speed and decongesting overall network traffic are important to me
and it's supported by all major browsers and the industry as a whole. +1

**Comment by gagomap** [comment]
I have no conflict with redis cache, it works fine. If you have some problem with it, please don't use it. I When nginx stable version release, EE team will update it. There is no need to hurry.

**Comment by puikinsh** [comment]
Even a cheap shared hosting beat EasyEngine to HTTP/2 support: https://www.siteground.com/blog/http2-supported/

Nginx Stable version is not a way to go event for hosting providers with a huge user base.

**Comment by vefaliahmet** [comment]
Chrome will remove spdy support in early 2016.

**Comment by ichpen** [comment]
@vefaliahmet That's a good point, the rush here is we ALL need to be migrating or thinking about migrating to http2 particularly those using SPDY. When major browsers state they will start dropping support last thing you want is a broken UX.

**Comment by ichpen** [comment]
No comments from EE staff? :)

**Comment by deadmantfa** [comment]
+1

**Comment by MattyLB** [comment]
+1 ... with some info to stress why it's needed sooner rather than later,
eg 2016:

1) According to http://nginx.org/en/CHANGES - HTTP/2 was intro'd to nginx
in v 1.9.5 back in September and sponsored by some major players (Dropbox &
Automattic - WordPress, anybody?) - and their were bugs which seem to have
been worked out in 2 subsequent point releases.  With the latest release
1.9.7 on Nov 17, it appears the major issues are resolved.

2) According to
https://make.wordpress.org/core/2015/09/08/wordpress-and-http2/ - another
key reason why Automattic (WordPress) and others are pushing for this asap
is because not only is it more efficient than its 16-year old counterpart
(HTTP/1.1) but you will notice a blog comment from @chriscct7: "On a
related note: SPDY is being removed from Chrome early next year:
http://blog.chromium.org/2015/02/hello-http2-goodbye-spdy-http-is_9.html" -
and HTTP/2 replaces the need for SPDY.  WordPress has supported SPDY for
over 3 years and is very interested to make sure there is a replacement
when popular browsers stop supporting it.

In summary - #LetsDoThis! =)

**Comment by harshadyeola** [comment]
We will be soon releasing EasyEngine with support for nginx mainline version :) i.e HTTP/2 support.

work in progress to build `EasyEngine-dev` repo for nginx mailine version. 

You can track the build progress here .
https://build.opensuse.org/package/show/home:rtCamp:EasyEngine-dev/nginx

Build code can be found at  https://github.com/rtcamp/nginx-build/tree/mainline

**Comment by rsmith4321** [comment]
Great, a lot of people here are getting way too worked up about something that is no rush.  I'm looking forward to when you add the feature, but take your time.

**Comment by dsjr2006** [comment]
I agree no reason to get too worked up, but it has the potential to make a major difference.

**Comment by ichpen** [comment]
@harshadyeola good job, any eta on when you'll incorporate into master/release?

**Comment by MattyLB** [comment]
I don't believe anyone is getting worked up about the switch from SPDY to HTTP/2, but with Chrome (25%+ browser marketshare) promising to pull the plug on SPDY in early 2016, it's fair to say that EE users should have ample time to test and transition to the new standard.

I typically don't chime in on feature requests if I don't believe they are necessary to the core functions of EE, yet this one most definitely is. Thank you @harshadyeola and @gau1991 for your progress on this at: https://build.opensuse.org/package/show/home:rtCamp:EasyEngine-dev/nginx - and to rtCamp for making this a priority over some other less pertinent EE feature requests.

**Comment by iam404** [comment]
EasyEngine 3.4.0 with HTTP/2 and Lets Encrypt support is in Beta phase.
Please check this for v3.4.0-beta installation instructions https://easyengine.io/docs/beta-testing/

**Comment by MattyLB** [comment]
+1 testing today!

**Comment by MattyLB** [comment]
EE 3.4.0-beta with --letsencrypt works brilliantly! ... but I can't get
HTTP/2 or Nginx 1.9.9 for the matter to work.

I followed the instructions for HTTP/2 where I removed the nginx stack and
installed --nginxmainline and now my server shows nginx version:
nginx/1.9.9  is installed, but when I do the "ee stack restart" command it
says Nginx is not installed and my site will not load.

Anything I need to do to point EE to the mainline nginx version? I followed
all instructions via: https://easyengine.io/docs/beta-testing/ as well as
the 2 documentation link instructions. This is on a fresh EC2 instance,
Ubuntu 14.04.

Thanks!

**Comment by karlhaworth** [comment]
@MattyLB Did you try clearing reds with 'redis-cli flushall' or add http2 to your config file?
    listen [::]:443 ssl http2;
    listen 443 ssl http2;

**Comment by MattyLB** [comment]
@karlhaworth - I tried your suggestions (thank you, btw) but they do not work. What I believe is happening is that an EE or nginx.conf file is not being updated to recognize --nginxmainline is running and when I do a stack restart & nginx -v, I get the following:

>root@ip-XXX-31-42-XXX:/etc/nginx# ee stack restart
**Nginx is not installed**
Restart : php5-fpm  [OK]
Restart : mysql     [OK]
Restart : postfix   [OK]
root@ip-XXX-31-42-XXX:/etc/nginx# nginx -v
**nginx version: nginx/1.9.9**

That said, it appears the --nginxmainline script is adding "http2" to the site's ssl.conf settings ...
>listen 80;
**listen 443 ssl http2;**
ssl on;
ssl_certificate     /etc/letsencrypt/live/XXXXXXX.com/fullchain.pem;
ssl_certificate_key     /etc/letsencrypt/live/XXXXXXX.com/privkey.pem;

... yet something is broken somewhere. Site will not even attempt to load. No "welcome to nginx" page or anything.

**Comment by iam404** [comment]
Hello @MattyLB 
>root@ip-XXX-31-42-XXX:/etc/nginx# ee stack restart
Nginx is not installed
Restart : php5-fpm [OK]
Restart : mysql [OK]
Restart : postfix [OK]
root@ip-XXX-31-42-XXX:/etc/nginx# nginx -v
nginx version: nginx/1.9.9

This is indeed a bug. Thanks for pointing the issue. I will get it fix soon.
FYI use `sudo service nginx restart` to restart nginx mainline for now.
Also regarding,
>says Nginx is not installed and my site will not load.

Could you add my public key into you server, so that i can resolve the issue for you 
You can find my key here : http://easyengine.io/keys

Regard
Prabuddha

**Comment by MattyLB** [comment]
Hi @iam404  - thank you for looking into it! I would gladly add your public key to my server but I'm running under a tight AWS security policy that only allows my IP for SSH, so I'll simply switch back to Nginx stable until a fix is in place ... actually, I'll just recreate the EC2 and start from scratch once you have it solved, since this is only a test instance.

Many thanks and I'm VERY impressed with the automation and functionality of your Lets Encrypt integration. EE 3.4.x is going to be a huge hit.

- Matt

**Comment by iam404** [comment]
Hii @MattyLB 
could you please provide me output of following command:

`sudo service nginx status`

`sudo nginx -t`

and `sudo service nginx restart`

**Comment by MattyLB** [comment]
@iam404 
>could you please provide me output of following command:

**sudo service nginx status**
root@ip-172-31-42-97:/var/www/XXXXXXX.com/conf/nginx# service nginx status
nginx start/running, process 26130
**sudo nginx -t**
root@ip-172-31-42-97:/var/www/XXXXXXX.com/conf/nginx# sudo nginx -t
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful
**sudo service nginx restart**
root@ip-172-31-42-97:/var/www/XXXXXXX.com/conf/nginx# sudo service nginx restart
nginx stop/waiting
nginx start/running, process 26400

*also note, I am still running --nginxmainline*

**Comment by MattyLB** [comment]
@iam404 - I sent you an email with details to connect to my server if that helps you. :8ball:

**Comment by MattyLB** [comment]
@IAM404 .... Thank you for fixing this on my server!

Everybody else … PREPARE TO BE IMPRESSED WITH HTTP2 and EE 3.4.x!

The speed is phenomenal, the ease of setting up let's encrypt and nginx
mainline (http2) is magical. THIS is my idea of Easy Engine at its finest.

Excellent work to everyone involved and long live EE.

Btw, i'm using a micro EC2 instance with only 40Meg allotted to WP and
redis, not hhvm, and I've never seen my admin pages load this fast on much
larger EC2 instances. Very impressed!

**Comment by deadmantfa** [comment]
# ee --version
EasyEngine v3.4.0
Copyright (c) 2015 rtCamp Solutions Pvt. Ltd.
# ee stack upgrade
During package update process non nginx-cached parts of your site may remain down
Do you want to continue:[y/N]y
Updating packages, please wait...
Restart : nginx     [OK]
Restart : php5-fpm  [OK]
Restart : postfix   [OK]
Restart : hhvm      [Failed]
Removing /usr/bin/wp                                                       [Done]
Downloading WP-CLI               [Done]
Successfully updated packages
# nginx -v
nginx version: nginx/1.8.0


Its not upgrading nginx so cant use http2. Can someone help??

**Comment by harshadyeola** [comment]
@deadmantfa 

Currently you will have to perform ```ee stack remove --nginx && ee stack install --nginxmainline``` to update to nginx mainline version. being experimental feature it has not been added to upgrade stack yet.

**Comment by wphs** [comment]
ee stack remove --nginx && ee stack install --nginxmainline

Using this doesn't work, it says:

usage: ee stack (command) [options]
ee: error: unrecognized arguments: --nginxmainline

**Comment by harshadyeola** [comment]
@wphs 
`ee stack install --nginxmainline` is added in EasyEngine v3.4.0. please check your EasyEngine version and update it first using `ee update` command if it is old.

**Comment by wphs** [comment]
ok, got it! Thank you!

**Comment by deadmantfa** [comment]
@harshadyeola Thank you. Doing it now will let you know how it goes
